### ISSUE TITLE
duckdb.dll not found in AzureAutomation when opening a connection

### ISSUE THREAD

**Issue by larsthelord** [dfinke/PSDuckDB#5]
When opening the connection for DataSource = :memory:, the duckdb.dll i not located when running this module in Azure Automation. 
Tested both for Powershell 5.1 and PowerShell 7.2

Throwing exception points to DuckDBConnectionStringBuilder:
![image](https://github.com/user-attachments/assets/4f24248e-2f3e-4311-8250-46e12026963b)

but the stack trace shows that the duckdb.dll cannot be found:
![image](https://github.com/user-attachments/assets/5578e134-f708-4a52-8db1-97fc244377f2)

I am unsure where the duckdb.dll originates from. I couldn't find it in the DuckDB.NET-repo either, 
Could you share some information here and i could try to help debugging this.

**Comment by dfinke** [comment]
Thanks @larsthelord for trying PSDuckDB and reporting thi.

Hmm, not sure. Could you share the script? You can also try `psduckdb` in the console.

`duckdb.dll` is here https://github.com/dfinke/PSDuckDB/tree/main/lib and should be local when you do the install-module

**Comment by larsthelord** [comment]
The script is as simple as it gets.

```
$ErrorActionPreference = "Stop"
Import-Module PsDuckDb
try{
    $db = New-DuckDBConnection
    $db.Open()
} catch {
    $err = $_
    Get-Error
}
```

I don't know if you're familiar with Azure Automation, but it's basically a kind of sandbox in Azure that lets you PowerShell scripts.

The modules are installed via a Module-handler in Azure
![image](https://github.com/user-attachments/assets/643745e7-350a-4c11-aaf2-244fd7d3c1a1)

![image](https://github.com/user-attachments/assets/0ec28b07-644d-435a-9a70-a507df2645d1)


I did find the duckdb.dll in the repo. But what is it compiled from?

**Comment by dfinke** [comment]
Yes familiar with Az automation. I have not tried PSDuckDB there. 

Are you running it on linux?

Here are details on it. I installed it from the nuget package.

https://duckdb.net/docs/getting-started.html

**Comment by larsthelord** [comment]
The $PsVersiontalbe in Automation Account says that it is running on Windows, but it is som kind of stripped down container thing. It doesn't allow you to do all of the usual stuff.

$PsVersionTable:
![image](https://github.com/user-attachments/assets/af04300c-7cf9-4a63-b63f-ac9d521ee5fd)


![image](https://github.com/user-attachments/assets/a0607c64-9325-49a4-91db-03a261e3b806)
from here (https://learn.microsoft.com/en-us/azure/automation/automation-runbook-execution)


I built the DuckDb.dlls for linux and Windows. 
Got it working in PowerShell in wsl when including the .so-file, but when running it in Automation Account it now throws a different error:

![image](https://github.com/user-attachments/assets/7eea1b9d-8b15-446c-b3ec-3edfd0bdfbbd)

So I don't think it is running on Linux.

I also tried adding the windows duckdb.dll and the libduckdb.so, but it just went back to the original exception

**Comment by larsthelord** [comment]
I've done some more reading. I see that the duckdb.dll is an unmanaged dll compiled from C. 
This might be the reason is won't work in Azure Automation, and that there is some restriction on unmanaged code.


And when it comes to if the Automation Account is Linux or Windows?
It has a weird naming convention for its folders that is a mix of Windows and Linux standard.

Here is cwd, and the ModulePath for PsDuckDb
![image](https://github.com/user-attachments/assets/83ebf074-5a26-4219-aa12-e73d4ab3666a)



When I get some time, I will try creating a simple unmanaged dll, then referencing it in another dll to see if the same problem appears. 
I will also have a look at putting the unmanaged dll into a combined dll to see if that makes a difference

**Comment by dfinke** [comment]
@larsthelord thanks for digging in!

**Comment by dfinke** [comment]
@larsthelord DuckDB update to a new version. I updated PSDuckDB.

It now works on my local PS 5.1. Give it a try.

**Comment by larsthelord** [comment]
Hi again! 

Sorry for the long wait. I have tried the new version on Automation Account in both 5.1 and 7.2, but the problem still persists, but now it seems even more severe? 

![image](https://github.com/user-attachments/assets/9f42a0d3-1970-4c40-aa3f-e753a2a3dec8)
The error is not now catchable, and just causes a halt.

7.2 halts on import .
5.1 halts on open connection

**Comment by dfinke** [comment]
Thanks for the update. Folks who did have an issue witn 5.1 are now able to work.

I don't use Az Automation. I don't know how to debug the issue. 

Hopefully something will pop up on my radar about.

**Comment by fraktallars** [comment]
I will try the compiling-trick. Compiling all .dlls as one, and see if that helps. (I've just been very busy lately) 
The module works fine on azure function, but that runs in it's own container, while Automation Account has some kind of shared container with other jobs. I guess this is the reason for system calls being limited, and the reason for the unmanaged duckdb.dll is causing problems.

**Comment by larsthelord** [comment]
I will try the compiling-trick. Compiling all .dlls as one, and see if that helps. (I've just been very busy lately) 
The module works fine on azure function, but that runs in it's own container, while Automation Account has some kind of shared container with other jobs. I guess this is the reason for system calls being limited, and the reason for the unmanaged duckdb.dll is causing problems.

**Comment by dfinke** [comment]
Interesting. No rush - looking forward to hearing what you find out

**Comment by larsthelord** [comment]
Had a chat with Microsoft Azure Support, and it seems that Automation account has restrictions when it comes to unmanaged code. 
This is as i thought the reason it won't accept the references. 

It is also not possible to repackage unmanaged code as a part of managed .dll
As of now, it seems that there is no way to utilize duckdb in Automation Account. The only alternative is using Azure functions, a VM or an actual machine.

**Comment by dfinke** [comment]
@larsthelord thanks for tracking all of that.

Yes, the core DuckDB is C++, unmanaged. I use DuckDB.NET, managed, which surfaces that as a ADO type interface.

That is too bad.
